File: src/main/java/eu/happybe/openapi/mysql/QueryQueue.java
package eu.happybe.openapi.mysql;

import eu.happybe.openapi.OpenAPI;
import io.gomint.scheduler.Task;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

public class QueryQueue {

    private static final Map<AsyncQuery, Consumer<AsyncQuery>> callbacks = new HashMap<>();

    public static void submitQuery(AsyncQuery query) {
        QueryQueue.submitQuery(query, null);
    }

    public static void submitQuery(AsyncQuery query, Consumer<AsyncQuery> callback) {
        query.host = DatabaseData.getHost();
        query.user = DatabaseData.getUser();
        query.password = DatabaseData.getPassword();

        Task task = OpenAPI.getInstance().scheduler().executeAsync(query);
        task.onComplete(query::onCompletion);

        if(callback != null) {
            QueryQueue.callbacks.put(query, callback);
        }
    }

    public static void activateCallback(AsyncQuery query) {
        if(!callbacks.containsKey(query)) {
            return;
        }

        Consumer<AsyncQuery> callback = QueryQueue.callbacks.get(query);
        callback.accept(query);
    }
}

File: src/main/java/eu/happybe/openapi/mysql/QueryQueue.kt
package eu.happybe.openapi.mysql

import cn.nukkit.Server
import eu.happybe.openapi.OpenAPI
import java.util.*
import java.util.function.Consumer

object QueryQueue {
    private val callbacks: MutableMap<AsyncQuery, Consumer<AsyncQuery>> = HashMap()
    fun submitQuery(query: AsyncQuery) {
        submitQuery(query, null)
    }

    fun submitQuery(query: AsyncQuery, callback: Consumer<AsyncQuery>?) {
        query.host = DatabaseData.getHost()
        query.user = DatabaseData.getUser()
        query.password = DatabaseData.getPassword()
        Server.getInstance().scheduler.scheduleAsyncTask(OpenAPI.getInstance(), query)
        if (callback != null) {
            callbacks[query] = callback
        }
    }

    fun activateCallback(query: AsyncQuery) {
        if (!callbacks.containsKey(query)) {
            return
        }
        val callback = callbacks[query]!!
        callback.accept(query)
    }
}